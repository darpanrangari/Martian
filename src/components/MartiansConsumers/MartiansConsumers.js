import React from 'react';
import {connect} from 'react-redux';
import Consumer from '../Consumer/Consumer'
import {fetchConsumerData, setCurrentConsumer, updateConsumer} from '../../actions/index'
import Modal from '../Modal/Modal';

import './styles.css'

class MartiansConsumers extends React.Component {

    constructor(props) {
        super(props);
        this.getConsumers = this.getConsumers.bind(this)
        this.state = {
            selectedConsumer: {},
            show: false,
            formData: {
                total_budget: ""
            },
            formErrors: {total_budget: ""},
            totalBudgetValid: false
        }
    }


    componentDidMount() {
        this.props.fetchConsumerData();
    }


    showModal = () => {
        this.setState({
            show: true
        })
    }

    hideModal = () => {
        this.setState({
            show: false,
            formErrors: {totalBudget: ''}
        })
    }

    openModal = (consumerData) => {
        this.props.setCurrentConsumer(consumerData)
        this.setState({
            formData: consumerData
        })
        this.showModal()
    }

    getConsumers = () => {
        let consumerComp = null;

        const {consumers} = this.props;
        console.log('getCosumer:', consumers);
        if (consumers) {
            consumerComp = (
                <Consumer
                    data={consumers}
                    openModal={this.openModal}
                />
            )

        }
        return consumerComp;
    }

    handleChange = (event) => {
        event.preventDefault()

        const {name, value} = event.target;

        this.setState({
            formData: {
                ...this.state.formData,
                [name]: value
            }
        })
        this.validateField(name, value)
    }

    validateField(fieldName, value) {
        let fieldValidationErrors = this.state.formErrors;
        let totalBudgetValid = this.state.totalBudgetValid;

        switch (fieldName) {
            case 'total_budget':
                totalBudgetValid = value !== "";
                fieldValidationErrors.total_budget = totalBudgetValid ? ((value > this.state.formData.budget_spent) ? totalBudgetValid : 'Total budget can not be less then budget spent') : 'Please Enter a valid value.';
                break;
            default:
                break;
        }
        this.setState({
            formErrors: fieldValidationErrors,
            totalBudgetValid
        }, this.validateForm);
    }

    validateForm() {
        this.setState({formValid: !!(this.state.total_budget)});
    }

    handleSubmit = (event) => {
        event.preventDefault();

        this.props.updateConsumer(this.state.formData);

        this.hideModal()
    }

    render() {
        const { show, formData, formValid, formErrors} = this.state;
        return (
            <div>
                <h2> Martian Consumers </h2>

                {this.getConsumers()}

                <Modal
                    isOpen={show}
                    isClose={this.hideModal}
                    submitFrom={this.handleSubmit}
                    modalHeader={formData.name}
                    formValid={formValid}
                >

                    <div className="form-element">

                        <p className='error'>{formErrors.total_budget}</p>

                        <label htmlFor="total_budget">Total Budget</label>

                        <input type="number" name="total_budget" onChange={this.handleChange}
                               value={this.state.formData.total_budget}/>

                    </div>
                </Modal>
            </div>
        );
    }
}

const mapStateToProps = state => {
    return {
        consumers: state.data.consumers
    };
};
const mapDispatchToProps = {
    fetchConsumerData,
    setCurrentConsumer,
    updateConsumer
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(MartiansConsumers);
