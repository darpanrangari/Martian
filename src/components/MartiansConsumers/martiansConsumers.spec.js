import React from 'react';
import {shallow} from 'enzyme';
import {MartiansConsumers} from './MartiansConsumers';

describe('test modal component', () => {

    let component = '',
        instance = ''

    beforeEach(() => {

        const props = {
            fetchConsumerData: jest.fn(),
            setCurrentConsumer: jest.fn(),
            updateConsumer: jest.fn(),
            consumers: [{
                id: 2,
                name: "Solar Firma",
                budget: 1123.2200,
                budget_spent: 451.3754,
                date_of_first_purchase: "2120-01-14"
            }]
        }
        component = shallow(<MartiansConsumers {...props}/>);

        instance = component.instance();

    });

    it('should render', () => {
        expect(component).toMatchSnapshot();
    });

    it('should check heading', () => {
        expect(component.find('h2').text()).toEqual(' Martian Consumers ');
    });

    it('should show modal when called', () => {
        instance.showModal();
        expect(component.state().show).toBeTruthy();
    });

    it('should hide modal when called and reset form error and total budget', () => {
        instance.hideModal();
        expect(component.state().show).toBeFalsy();
        expect(component.state().formErrors.totalBudget).toEqual("");
        expect(component.state().formData.total_budget).toEqual("");
    });

    it('should call handleChange', () => {
        let event = {
            target: {
                name: 'budget',
                value: 12345
            },
            preventDefault: jest.fn()
        }

        spyOn(instance, 'validateField');

        instance.handleChange(event);

        expect(event.preventDefault).toHaveBeenCalled();

        expect(instance.validateField).toHaveBeenCalledWith("budget", 12345);
    });


});
