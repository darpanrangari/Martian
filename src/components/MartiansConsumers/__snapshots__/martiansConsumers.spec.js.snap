// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test modal component should render 1`] = `ShallowWrapper {}`;
